src/System.Text.Json/tests/JsonNodeTests.JsonNumber.cs

```diff
@@ -0,0 +1,632 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Xunit;
+
+namespace System.Text.Json.Tests
+{
+    public static partial class JsonNodeTests
+    {
+        [Fact]
+        public static void TestDefaultCtor()
+        {
+            var jsonNumber = new JsonNumber();
+            Assert.Equal(0, jsonNumber.GetByte());
+            Assert.Equal(0, jsonNumber.GetInt16());
+            Assert.Equal(0, jsonNumber.GetInt32());
+            Assert.Equal(0, jsonNumber.GetInt64());
+            Assert.Equal(0, jsonNumber.GetSingle());
+            Assert.Equal(0, jsonNumber.GetDouble());
+            Assert.Equal(0, jsonNumber.GetSByte());
+            Assert.Equal((ushort)0, jsonNumber.GetUInt16());
+            Assert.Equal((uint)0, jsonNumber.GetUInt32());
+            Assert.Equal((ulong)0, jsonNumber.GetUInt64());
+            Assert.Equal(0, jsonNumber.GetDecimal());
+        }
+
+        [Theory]
+        [InlineData(17)]
+        [InlineData(byte.MinValue)]
+        [InlineData(byte.MaxValue)]
+        public static void TestByte(byte value)
+        {
+            // Default constructor:
+            var jsonNumber = new JsonNumber();
+            jsonNumber.SetByte(value);
+            Assert.Equal(value, jsonNumber.GetByte());
+            Assert.True(jsonNumber.TryGetByte(out byte result));
+            Assert.Equal(value, result);
+
+            // Numeric type constructor:
+            jsonNumber = new JsonNumber(value);
+            Assert.Equal(value, jsonNumber.GetByte());
+            Assert.True(jsonNumber.TryGetByte(out result));
+            Assert.Equal(value, result);
+
+            // Implicit cast:
+            jsonNumber = value;
+            Assert.Equal(value, jsonNumber.GetByte());
+            Assert.True(jsonNumber.TryGetByte(out result));
+            Assert.Equal(value, result);
+
+            // String constructor:
+            jsonNumber = new JsonNumber(value.ToString());
+            Assert.Equal(value, jsonNumber.GetByte());
+            Assert.True(jsonNumber.TryGetByte(out result));
+            Assert.Equal(value, result);
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(-17)]
+        [InlineData(17)]
+        [InlineData(short.MinValue)]
+        [InlineData(short.MaxValue)]
+        public static void TestShort(short value)
+        {
+            // Default constructor:
+            var jsonNumber = new JsonNumber();
+            jsonNumber.SetInt16(value);
+            Assert.Equal(value, jsonNumber.GetInt16());
+            Assert.True(jsonNumber.TryGetInt16(out short result));
+            Assert.Equal(value, result);
```

Comment:
All of these tests are the same thing repeated 4 times, repeated N times.  Consider factoring.

```C#
private delegate bool TryGetValue<T>(out T result);

private static void SimpleNumberTests<T>(
        T value,
        Func<JsonNumber, T> ctor,
        Action<JsonNumber, T> setter,
        Func<T, JsonNumber> getter, TryGetValue<T> tryGetter)
{
    JsonNumber number = new JsonNumber();
    setter(number, value);
    AssertValue(value, number, getter, tryGetter);

    number = ctor(value);
    AssertValue(value, number, getter, tryGetter);

    number = new JsonNumber(value.ToString());
    AssertValue(value, number, getter, tryGetter);
}

private static void AssertValue<T>(
        T value,
        JsonNumber number,
        Func<T, JsonNumber> getter,
        TryGetValue<T> tryGetter)
{
    Assert.Equal(value, getter(number));
    Assert.True(tryGetter(number, out T result));
    Assert.Equal(value, result);
}

[Theory]
[InlineData(0)]
[InlineData(-17)]
[InlineData(17)]
[InlineData(short.MinValue)]
[InlineData(short.MaxValue)]
public static void TestShort(short value)
{
    SimpleNumberTests(
        value,
        v => new JsonNumber(v),
        (number, v) => number.SetInt16(v),
        number => number.GetInt16(),
        (number, out v) =>number.TryGetValue(out v));
}
```

or something like that.

It'd lose the implicit cast, but you could build those to call the same ultimate helper.

Reply:
(Actually, you should be able to do the implicit cast as another Func<JsonNumber, T>, something like `v => v`, which isn't quite the identity lambda it looks :smile:)

Reply:
Thanks, that's exactly what I was missing!